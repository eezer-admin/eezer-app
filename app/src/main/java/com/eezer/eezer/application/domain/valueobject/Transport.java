package com.eezer.eezer.application.domain.valueobject;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Entity class representing a transport.
 */
public class Transport implements Serializable {

    private String transportId;
    private String driverId;
    private String vehicleId;
    private String passengerName;
    private String passengerPhone;
    private String gender;
    private String reason;
    private List<Coordinate> coordinates;
    private Float distance;
    private Date startedTime;
    private Date endedTime;
    private Long duration;
    private String deviceInfo;

    public Transport() {
        coordinates = new ArrayList<>();
        distance = 0f;
    }

    public String getTransportId() {
        return transportId;
    }

    public void setTransportId(String transportId) {
        this.transportId = transportId;
    }

    public String getDriverId() {
        return driverId;
    }

    public void setDriverId(String driverId) {
        this.driverId = driverId;
    }

    public String getVehicleId() {
        return vehicleId;
    }

    public void setVehicleId(String vehicleId) {
        this.vehicleId = vehicleId;
    }

    public String getPassengerName() {
        return passengerName;
    }

    public void setPassengerName(String passengerName) {
        this.passengerName = passengerName;
    }

    public String getPassengerPhone() {
        return passengerPhone;
    }

    public void setPassengerPhone(String passengerPhone) {
        this.passengerPhone = passengerPhone;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public String getReason() {
        return reason;
    }

    public void setReason(String reason) {
        this.reason = reason;
    }

    public Float getDistance() {
        return distance;
    }

    public void setDistance(Float distance) {
        this.distance = distance;
    }

    public List<Coordinate> getCoordinates() {
        return coordinates;
    }

    public void setCoordinates(List<Coordinate> coordinates) {
        this.coordinates = coordinates;
    }

    public Long getDuration() {
        return this.duration;
    }

    public void setDuration(Long duration) {
        this.duration = duration;
    }

    public String getStartedTime() {
        return startedTime.toString();
    }

    public Date getStartedTimeAsDate() {
        return this.startedTime;
    }

    public Date getEndedTimeAsDate() {
        return this.endedTime;
    }

    public void setStartedTime(Date startedTime) {
        this.startedTime = startedTime;
    }

    public String getDeviceInfo() {
        return deviceInfo;
    }

    public void setDeviceInfo(String deviceInfo) {
        this.deviceInfo = deviceInfo;
    }

    public String getEndedTime() {

        if (endedTime != null) {
            return endedTime.toString();
        } else {
            return null;
        }
    }

    public void setEndedTime(Date endedTime) {
        this.endedTime = endedTime;
    }
}
