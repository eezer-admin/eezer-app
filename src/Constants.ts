export const ERROR_CODES = {
  LOGIN_FAILED: 'login_failed',
  FAILED_GETTING_USER_TRANSPORTS: 'failed_getting_user_transports',
  FAILED_SYNCING_USER_TRANSPORTS: 'failed_syncing_user_transports',
};

export const STORAGE_KEYS = {
  USER: 'USER',
  TRANSPORT: 'TRANSPORT',
  TRANSPORT_LOG: 'TRANSPORT_LOG',
};

export const TRANSPORT_REASON = {
  DELIVERY: 'Delivery',
  CHECKUP: 'Checkup',
  OTHER: 'Other',
  ACCIDENT: 'Accident',
  SICKNESS: 'Sickness',
  MISC: 'Misc',
};
