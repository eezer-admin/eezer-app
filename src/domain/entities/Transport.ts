import {
  GetBackendTransport,
  NewTransportData,
  PostBackendTransport,
  TransportCoordinate,
} from '@interfaces/Transport';
import { formatDuration, uuid } from '@src/Utils';
import { format, parseISO } from 'date-fns';

export class Transport {
  public id: number | null;
  public identifier: string | null;
  public started: string | null; // Timestamp.
  public ended: string | null; // Timestamp.
  public durationSeconds: number;
  public distanceMeters: number;
  public distance: string | null;
  public reason: string | null;
  public coordinates: TransportCoordinate[];
  public vehicleId: number | null;

  constructor(data: NewTransportData = {}) {
    this.id = data.id || null;
    this.identifier = data.identifier || uuid();
    this.started = data.started || null;
    this.ended = data.ended || null;
    this.durationSeconds = data.durationSeconds || 0;
    this.distanceMeters = data.distanceMeters || 0;
    this.distance = data.distance || '0 m';
    this.reason = data.reason || null;
    this.coordinates = data.coordinates || [];
    this.vehicleId = data.vehicleId || null;
  }

  isOngoing(): boolean {
    return !this.ended && !!this.started;
  }

  isNotOngoing(): boolean {
    return !!this.ended || !this.started;
  }

  isEnded(): boolean {
    return !!this.ended;
  }

  isStarted(): boolean {
    return !!this.started;
  }

  isSynced(): boolean {
    return this.id !== null;
  }

  isNotSynced(): boolean {
    return !this.isSynced();
  }

  getReadableDuration(): string {
    if (this.started && this.ended) {
      return formatDuration(this.started, this.ended);
    }
    if (!this.started) {
      return '00:00';
    }

    return formatDuration(this.started, new Date().toISOString());
  }

  getReadableDistance(): string {
    if (this.distanceMeters < 1000) {
      return `${this.distanceMeters} m`;
    }

    return `${Math.round((this.distanceMeters / 1000) * 100) / 100} km`;
  }

  getReadableStartDate(): string {
    if (!this.started) {
      return '';
    }

    try {
      return format(parseISO(this.started), 'yyyy.MM.dd');
    } catch (err) {
      return '-';
    }
  }

  getStartDateAsDateFormat(): Date | null {
    if (!this.started) {
      return null;
    }

    return new Date(parseISO(this.started));
  }

  toApiFormat(): PostBackendTransport {
    return {
      id: this.id,
      vehicle_id: this.vehicleId,
      started_at: this.started || '',
      ended_at: this.ended || '',
      distance_meters: this.distanceMeters,
      reason: this.reason,
      coordinates: this.coordinates.map((coordinate: TransportCoordinate) => {
        return {
          latitude: coordinate.latitude,
          longitude: coordinate.longitude,
          altitude: coordinate.altitude,
          logged_at: coordinate.timestamp,
        };
      }),
    } as PostBackendTransport;
  }

  fromApiFormat(data: GetBackendTransport): Transport {
    this.id = data.id;
    this.started = data.started_at;
    this.ended = data.ended_at;
    this.distanceMeters = data.distance_meters;
    this.reason = data.reason;

    return this;
  }
}
