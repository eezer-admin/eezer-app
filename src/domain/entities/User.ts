import { UserData, Vehicle } from '@interfaces/User';

export class User {
  private id: number;
  private first_name: string | null;
  private last_name: string | null;
  private email: string;
  private phone: string | null;
  public access_token: string;
  private vehicles: Vehicle[];

  constructor(data: UserData) {
    this.id = data.id;
    this.first_name = data.first_name;
    this.last_name = data.last_name;
    this.email = data.email;
    this.phone = data.phone || null;
    this.access_token = data.access_token;
    this.vehicles = data.vehicles || [];
  }
}
