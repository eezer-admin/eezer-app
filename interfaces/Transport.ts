import { Transport } from '@src/domain/entities/Transport';

export type TransportData = {
  id: number | null;
  identifier: string | null;
  started: string | null; // Timestamp.
  ended: string | null; // Timestamp.
  durationSeconds: number | null;
  distanceMeters: number;
  distance: string | null;
  reason: string;
  coordinates: TransportCoordinate[];
  vehicle_id: number | null;
};

export type TransportCoordinate = {
  latitude: number;
  longitude: number;
  altitude: number;
  timestamp: string;
};

export type NewTransportData = {
  id?: number | null;
  identifier?: string | null;
  started?: string | null; // Timestamp.
  ended?: string | null; // Timestamp.
  durationSeconds?: number | null;
  distanceMeters?: number;
  distance?: string | null;
  reason?: string;
  coordinates?: TransportCoordinate[];
  vehicle_id?: number | null;
};

export type TransportContextData = {
  transport: Transport | null;
  setTransport(transport: Transport): Promise<void>;
  completeTransport(): Promise<void>;
};
