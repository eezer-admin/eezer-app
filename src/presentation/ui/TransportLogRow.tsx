import { Transport } from '@src/domain/entities/Transport';
import { useContext, useState } from 'react';
import { Text, TouchableOpacity, View } from 'react-native';

import { AuthContext } from '../../../contexts/authContext';
import { __ } from '../../../localization/Localization';
import Styles from '../../../styles/Styles';

export type TransportLogRowProps = {
  transport: Transport;
};

export default function TransportLogRow(props: TransportLogRowProps) {
  const auth = useContext(AuthContext);
  const [showDetails, setShowDetails] = useState(false);

  return (
    <View
      style={{
        borderBottomColor: Styles.colors.white,
        borderBottomWidth: 5,
      }}>
      <TouchableOpacity
        onPress={() => {
          setShowDetails(!showDetails);
        }}
        style={{
          backgroundColor: showDetails ? Styles.colors.grayDark : null,
          paddingVertical: Styles.margins.large,
          paddingHorizontal: Styles.container.paddingHorizontal,
          flex: 1,
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'space-between',
        }}>
        <Text style={{ width: '33%' }}>{props.transport.getReadableStartDate()}</Text>
        <Text style={{ width: '33%', textAlign: 'center' }}>{props.transport.getReadableDuration()}</Text>
        <Text style={{ width: '33%', textAlign: 'right' }}>{props.transport.isSynced() ? __('Yes') : __('No')}</Text>
      </TouchableOpacity>

      {showDetails ? (
        <View
          style={{
            paddingVertical: Styles.margins.large,
            paddingHorizontal: Styles.container.paddingHorizontal,
            backgroundColor: Styles.colors.gray,
            flex: 1,
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-between',
          }}>
          <View style={{ width: '33%' }}>
            <Text style={{ fontSize: Styles.fontSizes.small }}>{__('Driver')}</Text>
            <Text style={{ fontSize: Styles.fontSizes.small }}>
              {auth.user?.first_name} {auth.user?.last_name}
            </Text>
          </View>
          <View style={{ width: '33%' }}>
            <Text style={{ textAlign: 'center', fontSize: Styles.fontSizes.small }}>{__('Distance')}</Text>
            <Text style={{ textAlign: 'center', fontSize: Styles.fontSizes.small }}>
              {props.transport.getReadableDistance()}
            </Text>
          </View>
          <View style={{ width: '33%' }}>
            <Text style={{ textAlign: 'right', fontSize: Styles.fontSizes.small }}>{__('Cause')}</Text>
            <Text style={{ textAlign: 'right', fontSize: Styles.fontSizes.small }}>{props.transport.reason}</Text>
          </View>
        </View>
      ) : null}
    </View>
  );
}
