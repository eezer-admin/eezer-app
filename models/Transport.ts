import { format, parse } from 'date-fns'
import { TransportDriver, TransportInterface, UnmappedTransportInterface } from '../interfaces'

export class Transport implements TransportInterface {
    public id: String;
    public isUploaded: Boolean;
    public departsAt: Date;
    public distanceKm: Number;
    public secondsElapsed: Number;
    public cause: String;
    public driver: TransportDriver;

    constructor (data: UnmappedTransportInterface) {
      this.id = this.uuid()
      this.isUploaded = false
      this.departsAt = parse(data.date, 'yyyy-mm-dd', new Date())
      this.distanceKm = data.distance
      this.secondsElapsed = data.time
      this.cause = data.cause
      this.driver = {
        name: data.driverName
      }
    }

    getReadableDate (): String {
      return format(this.departsAt, 'yyyy.mm.dd')
    }

    uuid (): String {
      return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        const r = Math.random() * 16 | 0; const v = c == 'x' ? r : (r & 0x3 | 0x8)
        return v.toString(16)
      })
    }
}
