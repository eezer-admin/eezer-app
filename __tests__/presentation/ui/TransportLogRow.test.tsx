import { describe, expect, it } from '@jest/globals';
import renderer from 'react-test-renderer';

import TransportLogRow from '@presentation/ui/TransportLogRow';
import { TRANSPORT_REASON } from '@src/Constants';
import { Transport } from '@src/domain/entities/Transport';

describe('TransportLogRow', () => {
  it('renders correctly', () => {
    const transport = new Transport({ reason: TRANSPORT_REASON.ACCIDENT });
    const tree = renderer.create(<TransportLogRow transport={transport} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
