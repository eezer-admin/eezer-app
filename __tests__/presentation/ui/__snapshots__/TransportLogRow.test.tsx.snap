// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransportLogRow renders correctly 1`] = `
<View
  style={
    {
      "borderBottomColor": "white",
      "borderBottomWidth": 5,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": null,
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "opacity": 1,
        "paddingHorizontal": 15,
        "paddingVertical": 15,
      }
    }
  >
    <Text
      style={
        {
          "width": "33%",
        }
      }
    />
    <Text
      style={
        {
          "textAlign": "center",
          "width": "33%",
        }
      }
    >
      00:00
    </Text>
    <Text
      style={
        {
          "textAlign": "right",
          "width": "33%",
        }
      }
    >
      No
    </Text>
  </View>
</View>
`;
